### PR TITLE
Add support for client subscriptions

### PR DESCRIPTION
**Pull request by aashikam**
## Purpose

Fixes: https://github.com/ballerina-platform/ballerina-standard-library/issues/3560

## Pending
- [ ] Add test coverage
- [ ] Add spec to docs
- [ ] Add to changelog 

## Checklist
- [ ] Linked to an issue
- [ ] Updated the changelog
- [ ] Added tests
- [ ] Updated the spec
- [ ] Checked native-image compatibility
- [ ] No `commons` package changes (if there are any, please update the GraphQL version in [GraphQL tools](https://github.com/ballerina-platform/graphql-tools))

### CODE REVIEW

File: native/src/main/java/io/ballerina/stdlib/graphql/runtime/client/QueryExecutor.java [hunk 1]
```diff
@@ -37,6 +37,21 @@ public class QueryExecutor {
      */
     public static Object execute(Environment env, BObject client, BString document, Object variables,
                                  Object operationName, Object headers, BTypedesc targetType) {
+        if (targetType.getDescribingType().getTag() == PredefinedTypes.TYPE_STREAM.getTag()) {
+            @SuppressWarnings("UnckeckedCast")
```
Comment (MohamedSabthar) [inline native/src/main/java/io/ballerina/stdlib/graphql/runtime/client/QueryExecutor.java:41]: Do we need this annotation here ?

File: ballerina/client.bal [hunk 2]
```diff
@@ -15,25 +15,43 @@
...
 
 import ballerina/jballerina.java;
+import ballerina/uuid;
 import ballerina/http;
+import ballerina/websocket;
+
+const HTTP = "http://";
+const HTTPS = "https://";
+const WS = "ws://";
+const WSS = "wss://";
```
Comment (MohamedSabthar) [inline ballerina/client.bal:25]: We could move these to constants.bal file

File: ballerina/client.bal [hunk 3]
```diff
@@ -59,6 +77,52 @@ public isolated client class Client {
...
+        do {
+            lock {
+                if self.wsConnectionClosed {
+                    return;
+                }
+            }
+            lock {
+                foreach Subscriber subscriber in self.subscribersMap {
+                    check subscriber->unsubscribe();
+
```
Comment (DimuthuMadushan) [inline ballerina/client.bal:113]:
```suggestion
```

File: ballerina-tests/tests/35_client_test.bal [hunk 4]
```diff
@@ -733,23 +733,42 @@ isolated function testClientConfiguration() returns error? {
...
 
+
+@test:Config {
+    groups: ["client"]
+}
+isolated function testClientSubscription() returns error? {
+    string document = string `subscription { live { product { id } score } }`;
+    string url = "http://localhost:9090/reviews";
+    graphql:Client graphqlClient = check new (url);
+    stream<graphql:GenericResponseWithErrors|record {}|json> subscriptionStream = check graphqlClient->execute(document);
```
Comment (MohamedSabthar) [inline ballerina-tests/tests/35_client_test.bal:761]: Are we not supporting data binding in stream type?

File: ballerina/client.bal [hunk 5]
```diff
@@ -116,4 +160,113 @@ public isolated client class Client {
...
+    private isolated function initWsConnection() returns websocket:Error? {
+        lock {
+            if self.wsConnectionInit {
+                return;
+            }
+            websocket:Client wsClient = check self.getWebSocketClient();
+            ConnectionInitMessage message = {'type: WS_INIT};
+            check wsClient->writeMessage(message);
+            ConnectionAckMessage _ = check wsClient->readMessage();
```
Comment (DimuthuMadushan) [inline ballerina/client.bal:183]:
```suggestion
            _ = check wsClient->readMessage();
```

File: ballerina-tests/tests/35_client_test.bal (hunk 4 above)
Comment (MohamedSabthar) [reply ballerina-tests/tests/35_client_test.bal:761]: Ideally, we should support the following:
```ballerina
type LiveResponse record {
     record {
        record {
           string id;
        } product;
       int score;
     } live;
}

stream<LiveResponse, error> subscriptionStream = check cl->execute(doc);
```

File: ballerina/client.bal (hunk 2 above)
Comment (DimuthuMadushan) [inline ballerina/client.bal:25]: Can't we define these without the `://` part? IMO If it's possible, better to define that way.

File: ballerina/client.bal [hunk 6]
```diff
@@ -116,4 +160,113 @@ public isolated client class Client {
...
     }
+
+    private isolated function getWebSocketClient() returns websocket:Client|websocket:Error {
+        lock {
+            if self.wsClient is () {
+                self.wsConfig.subProtocols = [GRAPHQL_TRANSPORT_WS];
+                self.wsClient = check new (self.wsServiceUrl, self.wsConfig);
+                self.wsConnectionClosed = false;
+            }
+            return <websocket:Client> self.wsClient;
```
Comment (MohamedSabthar) [inline ballerina/client.bal:171]:
```suggestion
            return <websocket:Client>self.wsClient;
```

File: ballerina/client.bal [hunk 7]
```diff
@@ -116,4 +160,113 @@ public isolated client class Client {
...
+    }
+
+    private isolated function initWsConnection() returns websocket:Error? {
+        lock {
+            if self.wsConnectionInit {
+                return;
+            }
+            websocket:Client wsClient = check self.getWebSocketClient();
+            ConnectionInitMessage message = {'type: WS_INIT};
+            check wsClient->writeMessage(message);
```
Comment (MohamedSabthar) [inline ballerina/client.bal:182]:
```suggestion
            check wsClient->writeMessage({'type: WS_INIT});
```

File: ballerina/client.bal [hunk 8]
```diff
@@ -116,4 +160,113 @@ public isolated client class Client {
...
+            websocket:Client wsClient = check self.getWebSocketClient();
+            check self.initWsConnection();
+            string id = uuid:createType1AsString();
+            Subscriber subscriber = new (id, wsClient);
+            lock {
+                self.subscribersMap[id] = subscriber;
+            }
+            json payload = getGraphqlPayload(document, variables, operationName);
+            json graphqlPayload = {'type: WS_SUBSCRIBE, id, payload};
+            check wsClient->writeMessage(graphqlPayload);
```
Comment (MohamedSabthar) [inline ballerina/client.bal:201]:
```suggestion
            check wsClient->writeMessage({'type: WS_SUBSCRIBE, id, payload});
```

File: ballerina/subscriber.bal [hunk 9]
```diff
@@ -0,0 +1,77 @@
...
+        }
+    }
+    
+    isolated remote function unsubscribe() returns ClientError? {
+        lock {
+            if self.unsubscribed {
+                return;
+            }
+            CompleteMessage message = {'type: WS_COMPLETE, id: self.id};
+            websocket:Error? response = self.wsClient->writeMessage(message);
```
Comment (MohamedSabthar) [inline ballerina/subscriber.bal:70]:
```suggestion
            websocket:Error? response = self.wsClient->writeMessage({'type: WS_COMPLETE, id: self.id});
```

File: ballerina/subscriber.bal [hunk 10]
```diff
@@ -0,0 +1,77 @@
...
+    }
+
+    public isolated function next() returns record{|json value;|}? {
+        lock {
+            if self.unsubscribed || !self.wsClient.isOpen() {
+                return ();
+            }
+            while self.messages.length() == 0 {
+                runtime:sleep(1);
+            }
```
Comment (MohamedSabthar) [inline ballerina/subscriber.bal:44]: We may need to find a better way to block the calling thread. sleeping for 1sec may not be optimal

File: ballerina/subscriber.bal [hunk 11]
```diff
@@ -0,0 +1,77 @@
...
+            SubscriberMessage message = self.messages.shift();
+            if message is CompleteMessage {
+                self.unsubscribed = true;
+                return ();
+            }
+            json payload = message.payload;
+            return {value: payload.clone()};
```
Comment (MohamedSabthar) [inline ballerina/subscriber.bal:51]:
```suggestion
            return {value: message.payload.clone()};
```

File: ballerina/records.bal [hunk 12]
```diff
@@ -15,6 +15,7 @@
 // under the License.
 
 import ballerina/http;
+import ballerina/websocket;
 import graphql.parser;
```
Comment (MohamedSabthar) [inline ballerina/records.bal:19]:
```suggestion
import graphql.parser;

import ballerina/http;
import ballerina/websocket;
```

File: ballerina/client.bal [hunk 13]
```diff
@@ -15,25 +15,43 @@
 // under the License.
 
 import ballerina/jballerina.java;
+import ballerina/uuid;
 import ballerina/http;
+import ballerina/websocket;
```
Comment (MohamedSabthar) [inline ballerina/client.bal:20]:
```suggestion
import ballerina/http;
import ballerina/jballerina.java;
import ballerina/uuid;
import ballerina/websocket;
```

File: ballerina/subscriber.bal (hunk 10 above)
Comment (MohamedSabthar) [reply ballerina/subscriber.bal:44]: May be we can explore the option of maintaining a `LinkedBlockingQueue` in the native side

File: ballerina/subscriber.bal [hunk 14]
```diff
@@ -0,0 +1,77 @@
...
+
+    isolated function getStream() returns stream<GenericResponseWithErrors|record{}|json> {
+        stream<GenericResponseWithErrors|record{}|json> subscription = new (self);
+        return subscription;
+    }
+
+    public isolated function next() returns record{|json value;|}? {
+        lock {
+            if self.unsubscribed || !self.wsClient.isOpen() {
+                return ();
```
Comment (DimuthuMadushan) [inline ballerina/subscriber.bal:40]:
```suggestion
                return;
```

File: ballerina/subscriber.bal [hunk 15]
```diff
@@ -0,0 +1,77 @@
...
+            }
+            while self.messages.length() == 0 {
+                runtime:sleep(1);
+            }
+            SubscriberMessage message = self.messages.shift();
+            if message is CompleteMessage {
+                self.unsubscribed = true;
+                return ();
```
Comment (DimuthuMadushan) [inline ballerina/subscriber.bal:48]:
```suggestion
                return;
```

**Comment by ThisaruGuruge** [comment]
Let's park this for now as we need a [pipe](https://central.ballerina.io/xlibb/pipe)-like implementation to handle the subscriptions. (Using sleep is not ideal and may cause resource leaks). Implementing such functionality will take time and given the priorities, we decided to differ this for now. 

We can re-open this later if/when a requirement arises.

**Pull request closed without merging**